doc/writing_an_excmd.md

```diff
@@ -0,0 +1,88 @@
+
+WebExtensions are broken into two major components, the background script and the content script. You should read this page on MDN: [Anatomy of a WebExtension](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Anatomy_of_a_WebExtension).
+
+Tridactyl can run excmds in either context. In addition to extracting the signature for completion, [`excmd_macros.py`](../scripts/excmd_macros.py) splits [`excmds.ts`](../src/excmds.ts) into one version that runs in each tab's content script and one version that runs in Tridactyl's background script. Historically, you would annotate  your excmd to make sure it ended up in the right place. However, we've recently moved to a new system for making sure things happen in the right context, so you can just annotate your excmd's declaration with `//#both`:
+
+```typescript
+//#both
+export function yourexcmd() {}
+```
+
+### Documentation
+
+Tridactyl builds the `:help` page, and provides inline help when completing an excmd in the command-line, by running [TypeDoc](https://github.com/TypeStrong/typedoc) on [`excmds.ts`](../src/excmds.ts). You will need to add a documentation comment to the function which declares your excmd so that information about it appears on `:help`.
+
+### Implementation
+
+To help us move toward a future without hacks like [`excmd_macros.py`](../scripts/excmd_macros.py), we're currently trying to move all of the actual code out of `excmds.ts`. Additionally, as many WebExtension APIs are only available in either the background or content scripts but not both (e.g. [`tabs`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs) is background-only), you need some way to run your excmd in the right context.
+
+This mechanism is provided by the [`content_lib`](../src/content_lib) and [`background_lib`](../src/background_lib) directories. For each context-specific module in these directories, an "ambidextrous" module will be generated in [`lib/generated`](../src/lib/generated) that either invokes the original function (if in the correct context) or inserts a shim which automatically forwards invocations to the right context using [an appropriate messaging API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage)). You can put the implementation of your excmd in `background`, `content`, and `lib` and handle the details yourself if need or want to, but for the most part you should stick with the automatic forwarding.
+
+If your excmd fits into an established module (e.g. [`background_lib/tabs.ts`](../src/background_lib/tabs.ts)) you can put your implementation there. Otherwise, you can create a new module patterned after one of those. If necessary, import the "ambidextrous" wrapper module (the generated one in `src/lib/generated`) at the top of `excmds.ts`, in the portion marked `// Import excmd libs`. Then invoke the implementation in the declaration function you added earlier and you're done!
+
```

Comment:
Consider adding something like this:

`some older excmds use //#background or //#content in excmds.ts to create ambidextrous code. We're trying to move away from that, so please don't imitate this behavioour. If you have questions about those macros, ask on the chat channel or investigate excmds_macros.py`

If there is some reason other than that you got bored of porting all the code out (and it is totally fine to get bored of that), then please explain why some code still uses it.